Assets/HoloToolkit/UX/Scripts/Lines/LineRendererBase.cs

```diff
@@ -0,0 +1,173 @@
+        protected LineBase source;
+
+        [Header("Visual Settings")]
+        [Tooltip("Color gradient applied to line's normalized length")]
+        public Gradient LineColor;
+
+        public AnimationCurve LineWidth = AnimationCurve.Linear(0f, 0.05f, 1f, 0.05f);
+        [Range(0f, 10f)]
+        public float WidthMultiplier = 0.25f;
+
+        [Header ("Offsets")]
+        [Range(0f, 10f)]
+        [Tooltip("Normalized offset for color gradient")]
+        public float ColorOffset = 0f;
+        [Range(0f, 10f)]
+        [Tooltip("Normalized offset for width curve")]
+        public float WidthOffset = 0f;
+        [Range(0f, 10f)]
+        [Tooltip("Normalized offset for rotation offset")]
+        public float RotationOffset = 0f;
+
+        [Header ("Point Placement")]
+        [Tooltip("Method for gathering points along line. Interpolated uses normalized length. FromSource uses line's base points. (FromSource may not look right for all Line types.)")]
+        public StepModeEnum StepMode = StepModeEnum.Interpolated;
+        [Range(0, 2048)]
+        [Tooltip("Number of steps to interpolate along line in Interpolated step mode")]
+        [ShowIfEnumValue("StepMode", StepModeEnum.Interpolated)]
+        public int NumLineSteps = 10;
+
+        [FeatureInProgress]
+        public InterpolationModeEnum InterpolationMode = InterpolationModeEnum.FromLength;
+        [FeatureInProgress]
+        [Range(0.001f, 1f)]
+        public float StepLength = 0.05f;
+        [FeatureInProgress]
+        [Range(1, 2048)]
+        public int MaxLineSteps = 2048;
+        [FeatureInProgress]
+        public AnimationCurve StepLengthCurve = AnimationCurve.Linear(0f, 1f, 1f, 0.5f);
+
+        public virtual LineBase Target
+        {
+            get
+            {
+                return source;
+            }
+            set
+            {
+                source = value;
+            }
+        }
+
+        protected virtual Color GetColor (float normalizedLength)
+        {
+            if (LineColor == null)
+                LineColor = new Gradient();
+
+            return LineColor.Evaluate(Mathf.Repeat(normalizedLength + ColorOffset, 1f));
+        }
+
+        protected virtual float GetWidth(float normalizedLength)
+        {
+            if (LineWidth == null)
+                LineWidth = AnimationCurve.Linear(0f, 1f, 1f, 1f);
+
+            return LineWidth.Evaluate(Mathf.Repeat(normalizedLength + WidthOffset, 1f)) * WidthMultiplier;
+        }
+                      
+        protected virtual void OnEnable()
+        {
+            if (source == null)
+                source = gameObject.GetComponent<LineBase>();
+        }
+
+        private float[] normalizedLengths;
+
+#if UNITY_EDITOR
+        protected virtual void OnDrawGizmos()
+        {
+            if (Application.isPlaying)
+                return;
+            
+            if (source == null)
+                source = gameObject.GetComponent<LineBase>();
```

Comment:
you could even make source a property and get the component if it hasn't yet been initialized.

Also: braces

Reply:
Good idea.